File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/ui/fragments/AlarmSettingsFragment.java
/*
 * <!--
 *   ~ Copyright (c) 2017. ThanksMister LLC
 *   ~
 *   ~ Licensed under the Apache License, Version 2.0 (the "License");
 *   ~ you may not use this file except in compliance with the License. 
 *   ~ You may obtain a copy of the License at
 *   ~
 *   ~ http://www.apache.org/licenses/LICENSE-2.0
 *   ~
 *   ~ Unless required by applicable law or agreed to in writing, software distributed 
 *   ~ under the License is distributed on an "AS IS" BASIS, 
 *   ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *   ~ See the License for the specific language governing permissions and 
 *   ~ limitations under the License.
 *   -->
 */

package com.thanksmister.iot.mqtt.alarmpanel.ui.fragments;

import android.app.Dialog;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.preference.CheckBoxPreference;
import android.support.v7.preference.EditTextPreference;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceScreen;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import com.thanksmister.iot.mqtt.alarmpanel.BaseActivity;
import com.thanksmister.iot.mqtt.alarmpanel.R;
import com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions;
import com.thanksmister.iot.mqtt.alarmpanel.ui.Configuration;
import com.thanksmister.iot.mqtt.alarmpanel.ui.views.AlarmCodeView;
import com.thanksmister.iot.mqtt.alarmpanel.utils.DialogUtils;

import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_BROKER;
import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_CLIENT_ID;
import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_COMMAND_TOPIC;
import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_PASSWORD;
import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_PORT;
import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_STATE_TOPIC;
import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_TLS_CONNECTION;
import static com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions.PREF_USERNAME;
import static com.thanksmister.iot.mqtt.alarmpanel.ui.Configuration.PREF_ALARM_CODE;
import static com.thanksmister.iot.mqtt.alarmpanel.ui.Configuration.PREF_NOTIFICATIONS;
import static com.thanksmister.iot.mqtt.alarmpanel.ui.Configuration.PREF_PENDING_TIME;

public class AlarmSettingsFragment extends PreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

    private EditTextPreference brokerPreference;
    private EditTextPreference clientPreference;
    private EditTextPreference portPreference;
    private EditTextPreference commandTopicPreference;
    private EditTextPreference stateTopicPreference;
    private EditTextPreference userNamePreference;
    private EditTextPreference passwordPreference;
    private EditTextPreference pendingPreference;
    private CheckBoxPreference sslPreference;
    private CheckBoxPreference notificationsPreference;
    private Configuration configuration;
    private MQTTOptions mqttOptions;
    private Dialog alarmCodeDialog;
    private int defaultCode;
    private int tempCode;
    private boolean confirmCode = false;
  
    /**
     * Called during {@link #onCreate(Bundle)} to supply the preferences for this fragment.
     * Subclasses are expected to call {@link #setPreferenceScreen(PreferenceScreen)} either
     * directly or via helper methods such as {@link #addPreferencesFromResource(int)}.
     *
     * @param savedInstanceState If the fragment is being re-created from
     * a previous saved state, this is the state.
     * @param rootKey If non-null, this preference fragment should be rooted at the
     * {@link PreferenceScreen} with this key.
     */
    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.preferences);
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        
        super.onViewCreated(view, savedInstanceState);

        Preference buttonPreference = findPreference(PREF_ALARM_CODE);
        buttonPreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                showAlarmCodeDialog();
                return true;
            }
        });

        brokerPreference = (EditTextPreference) findPreference(PREF_BROKER);
        clientPreference = (EditTextPreference) findPreference(PREF_CLIENT_ID);
        portPreference = (EditTextPreference) findPreference(PREF_PORT);
        commandTopicPreference = (EditTextPreference) findPreference(PREF_COMMAND_TOPIC);
        stateTopicPreference = (EditTextPreference) findPreference(PREF_STATE_TOPIC);
        userNamePreference = (EditTextPreference) findPreference(PREF_USERNAME);
        passwordPreference = (EditTextPreference) findPreference(PREF_PASSWORD);
        pendingPreference = (EditTextPreference) findPreference(PREF_PENDING_TIME);
        sslPreference = (CheckBoxPreference) findPreference(PREF_TLS_CONNECTION);
        notificationsPreference = (CheckBoxPreference) findPreference(PREF_NOTIFICATIONS);
        
        if(isAdded()) {
            configuration = ((BaseActivity) getActivity()).getConfiguration();
            mqttOptions = ((BaseActivity) getActivity()).readMqttOptions();
        }
        
        brokerPreference.setText(mqttOptions.getBroker());
        clientPreference.setText(String.valueOf(mqttOptions.getClientId()));
        portPreference.setText(String.valueOf(mqttOptions.getPort()));
        commandTopicPreference.setText(mqttOptions.getCommandTopic());
        stateTopicPreference.setText(mqttOptions.getStateTopic());
        userNamePreference.setText(mqttOptions.getUsername());
        passwordPreference.setText(mqttOptions.getPassword());
        pendingPreference.setText(String.valueOf(configuration.getPendingTime()));
        sslPreference.setChecked(mqttOptions.getTlsConnection());
        notificationsPreference.setChecked(configuration.showNotifications());

        if(!TextUtils.isEmpty(mqttOptions.getBroker())) {
            brokerPreference.setSummary(mqttOptions.getBroker());
        }
        if(!TextUtils.isEmpty(mqttOptions.getClientId())) {
            clientPreference.setSummary(mqttOptions.getClientId());
        }
        if(!TextUtils.isEmpty(String.valueOf(mqttOptions.getPort()))) {
            portPreference.setSummary(String.valueOf(mqttOptions.getPort()));
        }
        if(!TextUtils.isEmpty(mqttOptions.getCommandTopic())) {
            commandTopicPreference.setSummary(mqttOptions.getCommandTopic());
        }
        if(!TextUtils.isEmpty(mqttOptions.getStateTopic())) {
            stateTopicPreference.setSummary(mqttOptions.getStateTopic());
        }
        if(!TextUtils.isEmpty(mqttOptions.getUsername())) {
            userNamePreference.setSummary(mqttOptions.getUsername());
        }
        if(!TextUtils.isEmpty(mqttOptions.getPassword())) {
            passwordPreference.setSummary(toStars(mqttOptions.getPassword()));
        }
        pendingPreference.setSummary(getString(R.string.preference_summary_pending_time, String.valueOf(configuration.getPendingTime())));
        
        // the first time we need to set the alarm code
        if(configuration.isFirstTime()) {
            showAlarmCodeDialog();
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        String value;
        switch (key) {
            case PREF_BROKER:
                value = brokerPreference.getText();
                if (!TextUtils.isEmpty(value)) {
                    mqttOptions.setBroker(value);
                    brokerPreference.setSummary(value);
                } else if(isAdded()) {
                    Toast.makeText(getActivity(), R.string.text_error_blank_entry, Toast.LENGTH_SHORT).show();
                }
                break;
            case PREF_CLIENT_ID:
                value = clientPreference.getText();
                if (!TextUtils.isEmpty(value)) {
                    mqttOptions.setClientId(value);
                    clientPreference.setSummary(value);
                } else if(isAdded()) {
                    Toast.makeText(getActivity(), R.string.text_error_blank_entry, Toast.LENGTH_LONG).show();
                    clientPreference.setText(mqttOptions.getClientId());
                }
                break;
            case PREF_PORT:
                value = portPreference.getText();
                if (value.matches("[0-9]+") && !TextUtils.isEmpty(value)) {
                    mqttOptions.setPort(Integer.valueOf(value));
                    portPreference.setSummary(String.valueOf(value));
                } else if(isAdded()) {
                    Toast.makeText(getActivity(), R.string.text_error_only_numbers, Toast.LENGTH_LONG).show();
                    portPreference.setText(String.valueOf(mqttOptions.getPort()));
                }
                break;
            case PREF_COMMAND_TOPIC:
                value = commandTopicPreference.getText();
                if (!TextUtils.isEmpty(value)) {
                    mqttOptions.setCommandTopic(value);
                    commandTopicPreference.setSummary(value);
                } else if(isAdded()) {
                    Toast.makeText(getActivity(), R.string.text_error_blank_entry, Toast.LENGTH_LONG).show();
                    commandTopicPreference.setText(mqttOptions.getCommandTopic());
                }
                break;
            case PREF_STATE_TOPIC:
                value = stateTopicPreference.getText();
                if (!TextUtils.isEmpty(value)) {
                    mqttOptions.setStateTopic(value);
                    stateTopicPreference.setSummary(value);
                }else if(isAdded()) {
                    Toast.makeText(getActivity(), R.string.text_error_blank_entry, Toast.LENGTH_LONG).show();
                    stateTopicPreference.setText(mqttOptions.getStateTopic());
                }
                break;
            case PREF_USERNAME:
                value = userNamePreference.getText();
                mqttOptions.setUsername(value);
                userNamePreference.setSummary(value);
                break;
            case PREF_PASSWORD:
                value = passwordPreference.getText();
                mqttOptions.setPassword(value);
                passwordPreference.setSummary(toStars(value));
                break;
            case PREF_PENDING_TIME:
                value = pendingPreference.getText();
                if (value.matches("[0-9]+") && !TextUtils.isEmpty(value)) {
                    int pendingTime = Integer.parseInt(value);
                    if(pendingTime < 10) {
                        if(isAdded()) {
                            ((BaseActivity) getActivity()).showAlertDialog(getString(R.string.text_error_pending_time_low));
                        }
                    }
                    configuration.setPendingTime(pendingTime);
                    pendingPreference.setText(String.valueOf(pendingTime));
                    pendingPreference.setSummary(getString(R.string.preference_summary_pending_time, String.valueOf(pendingTime)));
                } else if(isAdded()) {
                    Toast.makeText(getActivity(), R.string.text_error_only_numbers, Toast.LENGTH_LONG).show();
                    pendingPreference.setText(String.valueOf(configuration.getPendingTime()));
                }
                break;
            case PREF_TLS_CONNECTION:
                boolean checked = sslPreference.isChecked();
                mqttOptions.setTlsConnection(checked);
                break;
            case PREF_NOTIFICATIONS:
                boolean notify = notificationsPreference.isChecked();
                configuration.setNotifications(notify);
                break;
        }
    }

    private String toStars(String text) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            sb.append('*');
        }
        text = sb.toString();
        return text;
    }

    private void hideAlarmCodeDialog() {
        if(alarmCodeDialog != null) {
            alarmCodeDialog.dismiss();
            alarmCodeDialog = null;
        }
    }
    
    private void showAlarmCodeDialog() {
        // store the default alarm code
        defaultCode = configuration.getAlarmCode();
        hideAlarmCodeDialog();
        if(isAdded()) {
            alarmCodeDialog = DialogUtils.showCodeDialog(((BaseActivity) getActivity()), confirmCode, new AlarmCodeView.ViewListener() {
                @Override
                public void onComplete(int code) {
                    if (code == defaultCode) {
                        confirmCode = false;
                        hideAlarmCodeDialog();
                        Toast.makeText(getActivity(), R.string.toast_code_match, Toast.LENGTH_LONG).show();
                    } else if (!confirmCode) {
                        tempCode = code;
                        confirmCode = true;
                        hideAlarmCodeDialog();
                        showAlarmCodeDialog();
                    } else if (confirmCode && code == tempCode) {
                        configuration.setAlarmCode(tempCode);
                        configuration.setFirstTime(false);
                        tempCode = 0;
                        confirmCode = false;
                        hideAlarmCodeDialog();
                        Toast.makeText(getActivity(), R.string.toast_code_changed, Toast.LENGTH_LONG).show();
                    } else if (confirmCode) {
                        tempCode = 0;
                        confirmCode = false;
                        hideAlarmCodeDialog();
                        Toast.makeText(getActivity(), R.string.toast_code_not_match, Toast.LENGTH_LONG).show();
                    }
                }
                @Override
                public void onError() {
                }
                @Override
                public void onCancel() {
                    confirmCode = false;
                    hideAlarmCodeDialog();
                    Toast.makeText(getActivity(), R.string.toast_code_unchanged, Toast.LENGTH_SHORT).show();
                }
            }, new DialogInterface.OnCancelListener() {
                @Override
                public void onCancel(DialogInterface dialogInterface) {
                    confirmCode = false;
                    Toast.makeText(getActivity(), R.string.toast_code_unchanged, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }
}

File: app/src/main/java/com/thanksmister/iot/mqtt/alarmpanel/ui/fragments/AlarmSettingsFragment.kt
/*
 * Copyright (c) 2018 ThanksMister LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thanksmister.iot.mqtt.alarmpanel.ui.fragments

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.view.View
import androidx.preference.EditTextPreference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SwitchPreference
import com.thanksmister.iot.mqtt.alarmpanel.R
import com.thanksmister.iot.mqtt.alarmpanel.network.MQTTOptions
import com.thanksmister.iot.mqtt.alarmpanel.persistence.Configuration
import com.thanksmister.iot.mqtt.alarmpanel.ui.activities.SettingsActivity
import dagger.android.support.AndroidSupportInjection
import javax.inject.Inject

class AlarmSettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    @Inject
    lateinit var configuration: Configuration

    @Inject
    lateinit var mqttOptions: MQTTOptions

    private val remoteCodeSwitchPreference: SwitchPreference by lazy {
        findPreference(PREF_ALARM_REMOTE_CODE) as SwitchPreference
    }

    private val awayModePreference: SwitchPreference by lazy {
        findPreference("pref_mode_away") as SwitchPreference
    }

    private val homeModePreference: SwitchPreference by lazy {
        findPreference("pref_mode_home") as SwitchPreference
    }

    private val nightModePreference: SwitchPreference by lazy {
        findPreference("pref_mode_night") as SwitchPreference
    }

    private val customBypassModePreference: SwitchPreference by lazy {
        findPreference("pref_mode_custom_bypass") as SwitchPreference
    }

    private val pendingAwayPreference: EditTextPreference by lazy {
        findPreference(PREF_AWAY_PENDING_TIME) as EditTextPreference
    }

    private val pendingHomePreference: EditTextPreference by lazy {
        findPreference(PREF_HOME_PENDING_TIME) as EditTextPreference
    }

    private val pendingNightPreference: EditTextPreference by lazy {
        findPreference(PREF_NIGHT_PENDING_TIME) as EditTextPreference
    }

    private val pendingBypassPreference: EditTextPreference by lazy {
        findPreference(PREF_BYPASS_PENDING_TIME) as EditTextPreference
    }

    private val delayAwayPreference: EditTextPreference by lazy {
        findPreference(PREF_AWAY_DELAY_TIME) as EditTextPreference
    }

    private val delayHomePreference: EditTextPreference by lazy {
        findPreference(PREF_HOME_DELAY_TIME) as EditTextPreference
    }

    private val delayNightPreference: EditTextPreference by lazy {
        findPreference(PREF_NIGHT_DELAY_TIME) as EditTextPreference
    }

    private val delayBypassPreference: EditTextPreference by lazy {
        findPreference(PREF_BYPASS_DELAY_TIME) as EditTextPreference
    }

    override fun onAttach(context: Context) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
        (activity as SettingsActivity).supportActionBar?.setDisplayHomeAsUpEnabled(true)
        (activity as SettingsActivity).supportActionBar?.setDisplayShowHomeEnabled(true)
        (activity as SettingsActivity).supportActionBar?.title = (getString(R.string.preference_title_alarm))
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.alarm_preferences)
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        awayModePreference.isChecked = mqttOptions.alarmModeAway
        homeModePreference.isChecked = mqttOptions.alarmModeHome
        nightModePreference.isChecked = mqttOptions.alarmModeNight
        customBypassModePreference.isChecked = mqttOptions.alarmModeCustomBypass

        pendingAwayPreference.text = mqttOptions.pendingTimeAway.toString()
        pendingAwayPreference.summary = getString(R.string.pref_away_pending_summary, mqttOptions.pendingTimeAway.toString())
        pendingHomePreference.text = mqttOptions.pendingTimeHome.toString()
        pendingHomePreference.summary = getString(R.string.pref_home_pending_summary, mqttOptions.pendingTimeHome.toString())
        pendingNightPreference.text = mqttOptions.pendingTimeNight.toString()
        pendingNightPreference.summary = getString(R.string.preference_summary_night_pending_time, mqttOptions.pendingTimeNight.toString())
        pendingBypassPreference.text = mqttOptions.pendingTimeBypass.toString()
        pendingBypassPreference.summary = getString(R.string.pref_bypass_pending_summary, mqttOptions.pendingTimeBypass.toString())

        delayAwayPreference.text = mqttOptions.delayTimeAway.toString()
        delayAwayPreference.summary = getString(R.string.pref_away_delay_summary, mqttOptions.delayTimeAway.toString())
        delayHomePreference.text = mqttOptions.delayTimeHome.toString()
        delayHomePreference.summary = getString(R.string.pref_home_delay_summary, mqttOptions.delayTimeHome.toString())
        delayNightPreference.text = mqttOptions.delayTimeNight.toString()
        delayNightPreference.summary = getString(R.string.pref_night_delay_summary, mqttOptions.delayTimeNight.toString())
        delayBypassPreference.text = mqttOptions.delayTimeBypass.toString()
        delayBypassPreference.summary = getString(R.string.pref_bypass_delay_summary, mqttOptions.delayTimeBypass.toString())

        remoteCodeSwitchPreference.isChecked = mqttOptions.useRemoteCode
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
    }

    @SuppressLint("InlinedApi")
    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        val value: String
        when (key) {

            "pref_mode_home" -> {
                mqttOptions.alarmModeHome = homeModePreference.isChecked
            }
            "pref_mode_away" -> {
                mqttOptions.alarmModeAway = awayModePreference.isChecked
            }
            "pref_mode_night" -> {
                mqttOptions.alarmModeNight = nightModePreference.isChecked
            }
            "pref_mode_custom_bypass" -> {
                mqttOptions.alarmModeCustomBypass = customBypassModePreference.isChecked
            }
            PREF_AWAY_PENDING_TIME -> {
                mqttOptions.pendingTimeAway = pendingAwayPreference.text.toIntOrNull()?:60
                pendingAwayPreference.summary = getString(R.string.pref_away_pending_summary, mqttOptions.pendingTimeAway.toString())
            }
            PREF_HOME_PENDING_TIME -> {
                mqttOptions.pendingTimeHome = pendingHomePreference.text.toIntOrNull()?:60
                pendingHomePreference.summary = getString(R.string.pref_home_pending_summary, mqttOptions.pendingTimeHome.toString())
            }
            PREF_NIGHT_PENDING_TIME -> {
                mqttOptions.pendingTimeNight = pendingNightPreference.text.toIntOrNull()?:60
                pendingNightPreference.summary = getString(R.string.preference_summary_night_pending_time, mqttOptions.pendingTimeNight.toString())
            }
            PREF_BYPASS_PENDING_TIME -> {
                mqttOptions.pendingTimeBypass = pendingBypassPreference.text.toIntOrNull()?:60
                pendingBypassPreference.summary = getString(R.string.pref_bypass_pending_summary, mqttOptions.pendingTimeBypass.toString())
            }
            PREF_AWAY_DELAY_TIME -> {
                mqttOptions.delayTimeAway = delayAwayPreference.text.toIntOrNull()?:60
                delayAwayPreference.summary = getString(R.string.pref_away_delay_summary, mqttOptions.delayTimeAway.toString())
            }
            PREF_HOME_DELAY_TIME -> {
                mqttOptions.delayTimeHome = delayHomePreference.text.toIntOrNull()?:60
                delayHomePreference.summary = getString(R.string.pref_home_delay_summary, mqttOptions.delayTimeHome.toString())
            }
            PREF_NIGHT_DELAY_TIME -> {
                mqttOptions.delayTimeNight = delayHomePreference.text.toIntOrNull()?:60
                delayNightPreference.summary = getString(R.string.pref_night_delay_summary, mqttOptions.delayTimeNight.toString())
            }
            PREF_BYPASS_DELAY_TIME -> {
                mqttOptions.delayTimeBypass = delayHomePreference.text.toIntOrNull()?:60
                delayBypassPreference.summary = getString(R.string.pref_bypass_delay_summary, mqttOptions.delayTimeBypass.toString())
            }
            PREF_ALARM_REMOTE_CODE -> {
                val checked = remoteCodeSwitchPreference.isChecked
                checked.let {
                    mqttOptions.useRemoteCode = checked
                }
            }
        }
    }

    companion object {
        const val PREF_AWAY_PENDING_TIME = "settings_pref_away_pending_time"
        const val PREF_HOME_PENDING_TIME = "settings_pref_home_pending_time"
        const val PREF_NIGHT_PENDING_TIME = "settings_pref_night_pending_time"
        const val PREF_BYPASS_PENDING_TIME = "settings_pref_bypass_pending_time"
        const val PREF_AWAY_DELAY_TIME = "settings_pref_away_delay_time"
        const val PREF_HOME_DELAY_TIME = "settings_pref_home_delay_time"
        const val PREF_NIGHT_DELAY_TIME = "settings_pref_night_delay_time"
        const val PREF_BYPASS_DELAY_TIME = "settings_pref_bypass_delay_time"
        const val PREF_ALARM_REMOTE_CODE = "settings_alarm_remote_code"
    }
}